### ISSUE TITLE
Agents false starts

### ISSUE THREAD

**Issue by mikhail-yarosh** [JetBrains/teamcity-kubernetes-plugin#27]
i'm wondering what is going on: i have a build which could be started on a two agent-pools. 
The first one with regular VMs (always up). The second - with kubernetes agents.

During i'm triggering the build, pool id is set to one of them, in my case it's static pool with id 53:
`curl -s --basic --user username:password --request POST "https://tc.int/httpAuth/app/rest/buildQueue" --header "Content-Type: application/json" --header "Accept: application/json" --data '{"buildType": {"id": "TestBuildID"}, "agent": {"pool": {"id": 53}} }'`

in response:
`https://tc.int/httpAuth/app/rest/agents?locator=compatible:(build:(id:2677655))` 
only VMs are listed. 
The most interesting thing: if we have a queue of such builds - cloud agent going to rising up! 
There are no tasks for them, but for some reasons they're starting as much as it possible and, as we use "after first build" termination policy, these agents running up infinitely.

UI starts going same: manual choose of pool ignored, if there are no available (free to use) agents. But started k8s agents haven't any jobs and waiting for termination signal infinitely.

Looks strange, actually...

**Comment by ekoshkin** [comment]
@mikhail-yarosh i will take a look

**Comment by mikhail-yarosh** [comment]
@ekoshkin could it be a trouble with TeamCity scheduler itself?

**Comment by ekoshkin** [comment]
@mikhail-yarosh it might be
please enable debug logging and share cloud and teamcity-server.log with me

**Comment by mikhail-yarosh** [comment]
teamcity-agent.log (from k8s container, timestamp in UTC):
```
[2018-03-28 10:45:00,168]   INFO - l.directories.DirectoryMapImpl - Cleaning up old checkout directories. Default lifetime = 192 hour(s) 
[2018-03-28 10:45:00,379]   INFO - dDirectoryBasedCleanupRegistry - Removing files from /opt/buildAgent/temp /.old 
[2018-03-28 10:45:00,757]   INFO - buildServer.AGENT.registration - Server supports the following communication protocols: [polling] 
[2018-03-28 10:45:00,758]   INFO - buildServer.AGENT.registration - Trying to register on server using 'polling' protocol. 
[2018-03-28 10:45:01,064]   INFO - ldServer.AGENT.PollingProtocol - Start polling server for commands 
[2018-03-28 10:45:01,065]   INFO - buildServer.AGENT.registration - Registered on server with id 663 and authorization token '<TOKEN_HERE>' 
[2018-03-28 10:45:01,065]   INFO - buildServer.AGENT.registration - If this is the first time this agent registered on the server make sure it is authorized by administrator in the server web UI. 
[2018-03-28 10:45:01,066]   INFO -    jetbrains.buildServer.AGENT - Agent name was '', but server returned new name 'tc-agents-infrastructure-33ac446d-dxjr-1'. Agent name will be updated. 
[2018-03-28 10:45:01,391]   INFO - ldServer.AGENT.PollingProtocol - New command is received from server "testLocal" {id = 1} 
[2018-03-28 10:45:03,081]   INFO -    jetbrains.buildServer.AGENT - Updating agent parameters on the server: AgentDetails{Name='tc-agents-infrastructure-33ac446d-dxjr-1', AgentId=663, BuildId=null, AgentOwnAddress='null', AlternativeAddresses=[10.40.112.3], Port=9090, Version='46812', PluginsVersion='46812-md5-a340f141f98e6a99fc9eca79b490e337', AvailableRunners=[allureReportGeneratorRunner, Ant, cargo-deploy-runner, Duplicator, ftp-deploy-runner, gradle-runner, Inspection, JPS, Maven2, rake-runner, SBT, simpleRunner, smb-deploy-runner, ssh-deploy-runner, ssh-exec-runner], AvailableVcs=[tfs, cvs, jetbrains.git, mercurial, svn, perforce], AuthorizationToken='<TOKEN_HERE>', PingCode='BUVKsXVgQHts7ethKqyWvtcClLiSTD7S'} 

```
teamcity-clouds.log 
```
[2018-03-28 13:44:28,836]   INFO [8 Flush Queue 1] - .instances.StartInstanceAction - Starting cloud instance: profile 'web-test'{id=kube-4, projectId=_Root}, jetbrains.buildServer.clouds.kubernetes.KubeCloudImageImpl@725b3b0e, hash=Zu8zCrM5hd4DlCiuEd6j4E3bQSCrDFXc, reason=Automatic start: Compatible with queued build test-buildID(promotionId=2695425) 
[2018-03-28 13:44:29,392]   INFO [8 Flush Queue 1] - .server.impl.CloudEventsLogger - Cloud instance entered 'starting' state, profile 'web-test'{id=kube-4, projectId=_Root}, jetbrains.buildServer.clouds.kubernetes.CachingKubeCloudInstance@70f4723a 
[2018-03-28 13:44:29,392]   INFO [8 Flush Queue 1] - .server.impl.CloudEventsLogger - Cloud instance start succeeded: profile 'web-test'{id=kube-4, projectId=_Root}, jetbrains.buildServer.clouds.kubernetes.CachingKubeCloudInstance@290ff377 
[2018-03-28 13:44:41,136]   INFO [uled executor 4] - .server.impl.CloudEventsLogger - Cloud instance entered 'running' state, profile 'web-test'{id=kube-4, projectId=_Root}, jetbrains.buildServer.clouds.kubernetes.CachingKubeCloudInstance@70f4723a 
[2018-03-28 13:45:00,993]   INFO [o-8111-exec-250] - r.impl.DBCloudStateManagerImpl - Image: jetbrains.buildServer.clouds.kubernetes.KubeCloudImageImpl@725b3b0e, profile: profile 'web-test'{id=kube-4, projectId=_Root} was marked to CONTAIN agent 
[2018-03-28 13:45:00,994]   INFO [o-8111-exec-250] - .server.impl.CloudEventsLogger - Detected cloud agent "tc-agents-infrastructure-33ac446d-dxjr-1" {id=663}, profile 'web-test'{id=kube-4, projectId=_Root}, jetbrains.buildServer.clouds.kubernetes.CachingKubeCloudInstance@70f4723a 

```
teamcity-server.log
[2018-03-28 13:44:28,197]   INFO - ity.BuildQueuePriorityOrdering - New item Queued build {Build promotion {promotion id=2695425, configuration={id=test-buildIDID, internal id=bt3751}}, item id=2695425:1:53} with weight 0.00 inserted at the default position 33 in the end of the queue 
[2018-03-28 13:44:28,303]   INFO - s.buildServer.ACTIVITIES.AUDIT - build_add_to_queue: Build BUILD_PROMOTION{id=2695425} was added to queue by "'test.user' (test user) {id=360}" 
[2018-03-28 13:44:28,319]   INFO - tbrains.buildServer.ACTIVITIES - Build added to queue; Queued build {Build promotion {promotion id=2695425, configuration={id=test-buildIDID, internal id=bt3751}, branch=<default>, queued build id=2695425:1:53, queued, custom parameters={experiment_name=test123, env.test.env=rctests}}, triggered by "test user", agent pool id=53} 
[2018-03-28 13:45:16,316]   INFO -   jetbrains.buildServer.SERVER - Running build saved to DB: build id=2695425, promotion id=2695425 
[2018-03-28 13:45:16,316]   INFO - tbrains.buildServer.ACTIVITIES - Build started; test-buildID {id=test-buildIDID, internal id=bt3751} #127 {build id=2695425, promotion id=2695425, branch=<default>, history=false, agent="test-client-04.qa" {id=370}, triggered by "'test.user' (test user) {id=360}" (##userId='360' type='user'). Started 2018-03-28 13:45:16.252, running} 
[2018-03-28 13:45:16,332]   INFO -   jetbrains.buildServer.SERVER - Build removed from queue (started): 2695425 
[2018-03-28 13:46:25,848]   INFO -   jetbrains.buildServer.SERVER - Build interrupted (INTERRUPTED_BY_USER) with details: [comment: , userId:360]. Build: test-buildID {id=test-buildIDID, internal id=bt3751} #127 {build id=2695425, promotion id=2695425, branch=<default>, history=false, agent="test-client-04.qa" {id=370}, triggered by "'test.user' (test user) {id=360}" (##userId='360' type='user'), canceled=[comment: , userId:360]. Started 2018-03-28 13:45:17.274, running} 
[2018-03-28 13:46:29,661]   INFO - tbrains.buildServer.ACTIVITIES - Build stopped on agent: #127 {build id=2695425, buildTypeId=test-buildIDID} 
[2018-03-28 13:46:52,043]   INFO -   jetbrains.buildServer.SERVER - Build finished received for build: #127 {build id=2695425, buildTypeId=test-buildIDID} 
[2018-03-28 13:46:52,050]   INFO - ndexer.EventBasedBuildsIndexer - Indexing finishing build: #127 {build id=2695425, buildTypeId=test-buildIDID} 
[2018-03-28 13:46:52,072]   INFO - de.impl.history.DBBuildHistory - Created build history entry; build id: 2695425, finish date: 2018-03-28 13:46:52.019 
[2018-03-28 13:46:52,077]   INFO - s.buildServer.ACTIVITIES.AUDIT - build_cancel: Build "#127 {build id=2695425, buildTypeId=test-buildIDID}" was canceled by "'test.user' (test user) {id=360}" 
[2018-03-28 13:46:52,114]   INFO - tbrains.buildServer.ACTIVITIES - Finished build test-buildID {id=test-buildIDID, internal id=bt3751} #127 {build id=2695425, promotion id=2695425, branch=<default>, history=false, agent="test-client-04.qa" {id=370}, triggered by "'test.user' (test user) {id=360}" (##userId='360' type='user'), canceled=[comment: , userId:360]. Started 2018-03-28 13:45:17.274, running} 

As you can see above, the pool id=53 and it consists of the nodes with name test-client-*.qa ONLY.

But:
https://tc.int/httpAuth/app/rest/agents?locator=compatible:(build:(id:2695425))
```
<agents count="7" href="/httpAuth/app/rest/agents?locator=compatible:(build:(id:2695425))">
<agent id="351" name="test-client-02.qa" typeId="351" href="/httpAuth/app/rest/agents/id:351"/>
<agent id="365" name="test-client-03.qa" typeId="365" href="/httpAuth/app/rest/agents/id:365"/>
<agent id="370" name="test-client-04.qa" typeId="370" href="/httpAuth/app/rest/agents/id:370"/>
<agent id="371" name="test-client-06.qa" typeId="371" href="/httpAuth/app/rest/agents/id:371"/>
<agent id="372" name="test-client-05.qa" typeId="372" href="/httpAuth/app/rest/agents/id:372"/>
<agent id="373" name="test-client-01.qa" typeId="373" href="/httpAuth/app/rest/agents/id:373"/>
<agent id="663" name="tc-agents-infrastructure-33ac446d-dxjr-1" typeId="436" href="/httpAuth/app/rest/agents/id:663"/>
</agents>
```

And what is more interesting, for old build:
https://tc.int/httpAuth/app/rest/agents?locator=compatible:(build:(id:2677655))
```
<agents count="7" href="/httpAuth/app/rest/agents?locator=compatible:(build:(id:2677655))">
<agent id="351" name="test-client-02.qa" typeId="351" href="/httpAuth/app/rest/agents/id:351"/>
<agent id="365" name="test-client-03.qa" typeId="365" href="/httpAuth/app/rest/agents/id:365"/>
<agent id="370" name="test-client-04.qa" typeId="370" href="/httpAuth/app/rest/agents/id:370"/>
<agent id="371" name="test-client-06.qa" typeId="371" href="/httpAuth/app/rest/agents/id:371"/>
<agent id="372" name="test-client-05.qa" typeId="372" href="/httpAuth/app/rest/agents/id:372"/>
<agent id="373" name="test-client-01.qa" typeId="373" href="/httpAuth/app/rest/agents/id:373"/>
<agent id="663" name="tc-agents-infrastructure-33ac446d-dxjr-1" typeId="436" href="/httpAuth/app/rest/agents/id:663"/>
</agents>
```

So, @ekoshkin, what will you say?

**Comment by mikhail-yarosh** [comment]
@ekoshkin could you provide any information about issue?

**Comment by ekoshkin** [comment]
@mikhail-yarosh you are probably right, it looks like a bug in scheduler
are all the VM's busy when those builds are being triggered?
what are CPU ranks (agent parameter) of VM agent and k8s agent?

**Comment by mikhail-yarosh** [comment]
@ekoshkin 
> are all the VM's busy when those builds are being triggered?

not necessary

> what are CPU ranks (agent parameter) of VM agent and k8s agent? 

the VMs agents CPU rank: 523 
the agents CPU rank: 494

**Comment by ekoshkin** [comment]
@mikhail-yarosh please confirm that i understood you
You are saying that when you for example do custom build (Run...) and choose static pool with VM based agents for your build you got this build queued and K8s based agent is starting (even if some VM agents are idle). Your build after all starts on VM based agent.

**Comment by mikhail-yarosh** [comment]
@ekoshkin queueing is necessary. Tested one more time: it happens only when no any idle VM based agents. But when cloud agent runs, no one job for it.

**Comment by ekoshkin** [comment]
@mikhail-yarosh ок, got it. So the issue is - when you queue a build with some pool specified and all agents in that pool are busy you see another agent started in another compatible pool. Right?

**Comment by mikhail-yarosh** [comment]
@ekoshkin absolutely.

**Comment by ekoshkin** [comment]
I've created an issue in main TeamCity tracker https://youtrack.jetbrains.com/issue/TW-54485
What version of TeamCity do you use?

**Comment by mikhail-yarosh** [comment]
TeamCity Enterprise 2017.1.2 (build 46812)

**Comment by ekoshkin** [comment]
fixed in 2017.2.4 version of teamcity server

**Comment by ekoshkin** [comment]
@mikhail-yarosh in a week or so
